module/Finna/src/Finna/Recommend/Ontology.php

```diff
@@ -0,0 +1,512 @@
+                || $this->resultTotal >= $this->minLargeResultTotal)
+            ) {
+                // The Finto result is not a non-descriptor and there is a
+                // large number of search results so we will try to make an
+                // additional Finto call to see if there are narrower
+                // concepts.
+                if (!$this->canMakeApiCall()) {
+                    return;
+                }
+                if ($hyponymResults = $this->finto->narrower(
+                    $fintoResult['vocab'], $fintoResult['uri'],
+                    $fintoResult['lang'], true
+                )
+                ) {
+                    $termUri = $fintoResult['uri'] ?? null;
+                    foreach ($hyponymResults['narrower'] as $hyponymResult) {
+                        $this->addOntologyResult(
+                            $hyponymResult, self::TYPE_HYPONYM, $term, $termUri
+                        );
+                    }
+                }
+                $this->apiCallTotal += 1;
+            }
+        }
+    }
+
+    /**
+     * Adds an ontology result to the specified array.
+     *
+     * @param array       $fintoResult Finto result
+     * @param string      $resultType  Result type
+     * @param string      $term        The term searched for
+     * @param string|null $termUri     URI of the searched term if applicable
+     *
+     * @return void
+     */
+    protected function addOntologyResult(
+        $fintoResult, $resultType, $term, $termUri = null
+    ) {
+        $this->ontologyResultTotal += 1;
+
+        // Recommendation memory cookie key and value.
+        $key = $this->ontologyResultTotal;
+        if ($this->searchId) {
+            $key = $this->searchId . '-' . $key;
+        }
+        $value = $this->recommendationMemory->getDataString(
+            'Ontology', $fintoResult['prefLabel'], $term, $resultType
+        );
+
+        // Recommendation link.
+        $base = $this->urlHelper->__invoke('search-results');
+        $uriField = 'topic_uri_str_mv:' . $fintoResult['uri'];
+        $recommendedTerm = $fintoResult['prefLabel'];
+        if (preg_match('/\s/', $recommendedTerm)) {
+            $recommendedTerm = '"' . $recommendedTerm . '"';
+        }
+        $replace = $uriField . ' ' . $recommendedTerm;
+        $recommend = str_replace($term, $replace, $this->lookfor);
+        if ($termUri) {
+            $recommend = str_replace('topic_uri_str_mv:' . $termUri, '', $recommend);
+            $recommend = preg_replace('/\s+/', ' ', $recommend);
+        }
+        $params = [
+            'lookfor' => $recommend,
+            RecommendationMemory::PARAMETER_NAME => $key
+        ];
+        $params = array_merge($this->request->toArray(), $params);
+        unset(
+            $params['mod'], $params['params'], $params['searchId'],
+            $params['resultTotal']
+        );
+        $href = $base . '?' . http_build_query($params);
```

Comment:
Käytä ennemmin url helperiä koko url:n tekemiseen: `$this->url->__invoke('search-results', [], ['query' => $params]);`

Reply:
👍